smac/epm/partial_sparse_gaussian_process.py

```diff
@@ -0,0 +1,874 @@
+
+        Parameters
+        ----------
+        in_x: Optional[np.ndarray (N_in, D)]
+            Input data points. The dimensionality of X is (N, D),
+            with N as the number of points and D is the number of features.
+        in_y: Optional[np.ndarray (N,)]
+            The corresponding target values.
+        out_x: Optional[np.ndarray (N_out, D)]
+            If set to true the hyperparameters are optimized otherwise
+            the default hyperparameters of the kernel are used.
+        out_y: typing.Optional[np.ndarray (N_out)] = None
+        """
+        if in_x is None:
+            return None
+
+        in_x = torch.from_numpy(in_x)
+        in_y = torch.from_numpy(in_y)
+        if out_x is None:
+            self.gp_model = ExactGPModel(in_x, in_y, likelihood=self.likelihood, base_covar_kernel=self.kernel).double()
+        else:
+            out_x = torch.from_numpy(out_x)
+            out_y = torch.from_numpy(out_y)
+
+            if self.num_inducing_points <= in_y.shape[0]:
+                weights = torch.ones(in_y.shape[0]) / in_y.shape[0]
+                inducing_points = in_x[torch.multinomial(weights, self.num_inducing_points)]
+            else:
+                weights = torch.ones(out_y.shape[0]) / out_y.shape[0]
+                inducing_points = out_x[torch.multinomial(weights, self.num_inducing_points - in_y.shape[0])]
+                inducing_points = torch.cat([inducing_points, in_x])
+            self.gp_model = PartailSparseGPModel(in_x, in_y, out_x, out_y,
+                                                 likelihood=self.likelihood,
+                                                 base_covar_kernel=self.kernel,
+                                                 inducing_points=inducing_points).double()
+
+        mll = ExactMarginalLogLikelihood(self.likelihood, self.gp_model)
+        mll.double()
+        return mll
+
+    def _optimize(self) -> np.ndarray:
+        """
+        Optimizes the marginal log likelihood and returns the best found
+        hyperparameter configuration theta.
+
+        Returns
+        -------
+        theta : np.ndarray(H)
+            Hyperparameter vector that maximizes the marginal log likelihood
+        """
+        if isinstance(self.gp_model, PartailSparseGPModel):
+            self.gp_model.deactivate_inducing_points_grad()
+
+        x0, property_dict, bounds = module_to_array(module=self.gp)
+
+        self.property_dict = property_dict
+
+        bounds = np.asarray(bounds).transpose().tolist()
+        p0 = [x0]
+        n_tries = 5000
+        for i in range(n_tries):
+            try:
+                self.gp.pyro_sample_from_prior()
+                sample, _, _ = module_to_array(module=self.gp)
+                p0.append(sample.astype(np.float64))
+            except Exception as e:
+                if i == n_tries - 1:
+                    raise e
+                continue
```

Comment:
Please try to be more specific in what exceptions are okay to skip and add some documentation.